src/Common/src/System/Dynamic/Utils/CollectionExtensions.cs

```diff
@@ -57,24 +57,29 @@ public static int ListHashCode<T>(this IEnumerable<T> list)
         }
 
         [Pure]
-        public static bool ListEquals<T>(this ICollection<T> first, ICollection<T> second)
+        public static bool ListEquals<T>(this ReadOnlyCollection<T> first, ReadOnlyCollection<T> second)
         {
-            if (first.Count != second.Count)
+            if (first == second)
+            {
+                return true;
+            }
```

Comment:
It's hard to say considering it gets called into by a public member, but both being a singleton empty collection would be common, and if a CallInfo was created based on the members of another then it would.